src/Components/Server/src/Circuits/CircuitHost.cs

```diff
@@ -110,17 +110,19 @@ public Task InitializeAsync(ProtectedPrerenderComponentApplicationStore store, C
                         _options.RootComponents.JSComponents,
                         JSRuntime.ReadJsonSerializerOptions());
 
-                    // We add the root components *after* the circuit is flagged as open.
-                    // That's because AddComponentAsync waits for quiescence, which can take
-                    // arbitrarily long. In the meantime we might need to be receiving and
-                    // processing incoming JSInterop calls or similar.
+                    // Here, we add each root component but don't await the returned tasks so that the
+                    // components can be processed in parallel.
                     var count = Descriptors.Count;
+                    var pendingRenders = new Task[count];
```

Comment:
If you'd like this to avoid this allocation, you could use an array pool

```C#
var pendingRenders = ArrayPool<Task>.Shared.Rent(count);
try
{
  ...
   await Task.WhenAll(tasks.Take(count));
}
finally
{
   ArrayPool<Task>.Shared.Return(tasks);
}

Reply:
Is that really worth it though? This only happens once per circuit startup, and circuit startup involves (probably) hundreds of allocations. Do we really want to have to reason about the correctness of pooling logic?

Reply:
> Is that really worth it though?

Kinda why I didn't insist on it. It's a nice to have, but I wouldn't think this is critical especially in the absence of profiling data.